cadc-pkg-server/src/main/java/org/opencadc/pkg/server/PackageWriter.java

```diff
@@ -109,49 +115,118 @@ public void close() throws IOException {
+     *
+     * @param packageItem - item to be written to archive
+     * @throws IOException
+     * @throws InterruptedException
+     * @throws ResourceNotFoundException
+     * @throws TransientException
+     * @throws ResourceAlreadyExistsException
      */
     public void write(PackageItem packageItem) throws IOException, InterruptedException,
         ResourceNotFoundException, TransientException, ResourceAlreadyExistsException {
 
-        boolean openEntry = false;
+        // openEntry is set to true after an archive entry header
+        // is successfully written, in either writeFileContentEntry
+        // or write HttpContentEntry.  It's so the stream can be closed
+        // correctly in the 'finally'
+        this.openEntry = false;
 
         try {
-            // HEAD to get entry metadata
             // Implementations of PackageRunner that build the PackageItem list
             // should ensure that files exist before submitting as part of the
             // Iterator<PackageItem>
             URL packageURL = packageItem.getURL();
-            HttpGet get = new HttpGet(packageURL, true);
-            get.prepare();
-
-            // get information common to all entries
-            long contentLength = get.getContentLength();
-            Date lastModified = get.getLastModified();
-
-            // create entry (metadata) to be put to archive stream
-            log.debug("next package entry: " + packageItem.getRelativePath() + "," + contentLength + "," + lastModified);
-            ArchiveEntry e = createEntry(packageItem.getRelativePath(), contentLength, lastModified);
-
-            // put archive entry to stream
-            aout.putArchiveEntry(e);
 
-            // headers for entry have been written, body has not,
-            // so consider this entry 'open'
-            openEntry = true;
-
-            // Copy the get InputStream to the package OutputStream
-            // this is 'writing' the content of the file
-            InputStream getIOStream = get.getInputStream();
-            MultiBufferIO multiBufferIO = new MultiBufferIO();
-            multiBufferIO.copy(getIOStream, aout);
+            if (packageURL.getProtocol().equals("file")) {
+                writeFileContentEntry(packageURL, packageItem.getRelativePath());
+            } else {
+                writeHttpContentEntry(packageURL, packageItem.getRelativePath());
+            }
 
         } finally {
             if (openEntry) {
                 aout.closeArchiveEntry();
             }
         }
     }
+
+
+    /**
+     * Set up header for ArchiveEntry in the current ArchiveOutputStream.
+     * @param relativePath - relative path of item in the package file
+     * @param contentLength - (long) - bytes to be written
+     * @param lastModified - (Date) - last modified date of item
+     * @return
+     * @throws IOException
+     */
+    private boolean setArchiveEntryHeader(String relativePath, long contentLength, Date lastModified)
```

Comment:
Is the return type 'boolean' here necessary?  It only returns 'true' (never false) so that is the same as saying "this method completed", so don't think it adds value.

Reply:
It is relevant if the step for setting up the archive metadata fails, as compared to when writing the body of the archive entry fails.  - 'openEntry = true' means that the archive metadata write was successful. 'openEntry=false' means it wasn't.

Reply:
But since this cannot return false it should not return anything.

Reply:
ok